### ISSUE TITLE
CBS: all videos failing

### ISSUE THREAD

**Issue by 4cq2** [ytdl-org/youtube-dl#31864]
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

using an `assetType` of `Downloadable` (not DRM):

https://cbs.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_

~~~
[debug] Command-line args: ['--verbose', 'cbs:YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, pref cp1252
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.10.4 (CPython) - Windows-10-10.0.18363-SP0
[debug] exe versions: ffmpeg 2022-06-09-git-5d5a014199-essentials_build-www.gyan.dev
[debug] Proxy map: {}
[CBS] YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_: Downloading XML
ERROR: YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_: Failed to parse XML  (caused by
ParseError('no element found: line 1, column 0'))
~~~

same with `assetType` of `StreamPack` (not DRM):

<https://paramountplus.com/movies/video/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_>

~~~
[debug] Command-line args: ['--verbose', 'cbs:wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, pref cp1252
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.10.4 (CPython) - Windows-10-10.0.18363-SP0
[debug] exe versions: ffmpeg 2022-06-09-git-5d5a014199-essentials_build-www.gyan.dev
[debug] Proxy map: {}
[CBS] wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_: Downloading XML
ERROR: wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_: Failed to parse XML  (caused by
ParseError('no element found: line 1, column 0'))
~~~

## Description

looks like the issue is that these type URLs are now returning empty responses:

<http://can.cbs.com/thunder/player/videoPlayerService.php?partner=cbs&contentId=wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_>

I believe this started in the last day or two, not sure if temporary or permanent issue.

**Comment by dirkf** [comment]
As you haven't shown a verbose log, it's difficult to diagnose: clearly you aren't running the release build or a new git version, as neither recognise the CBS URL.

However the same result occurs with yt-dlp.

**Comment by 4cq2** [comment]
> As you haven't shown a verbose log, it's difficult to diagnose: clearly you aren't running the release build or a new git version, as neither recognise the CBS URL.

I did show a verbose log, what are you talking about? I ran with `--verbose` as instructed. Also if you look closely, you'll notice I didn't provide a URL, I provided the `cbs:ID`

**Comment by dirkf** [comment]
Quite so, late at night in the UK, but you did omit the backtrace, so that's my weak excuse.

Using the `cbs:ID`, all three versions of interest (release yt-dl and yt-dlp, yt-dl git master) fail in the same way, as in the posted logs.

Someone (OP perhaps?) will have to do some research, eg on how this works in region in a browser, to fix this.

**Comment by october262** [comment]
the same video at paramount +  https://www.paramountplus.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_/  works with yt-dlp & ytdl-patched but does not work with youtube-dl

**Comment by dirkf** [comment]
Looks like this is all blocked (403) from UK.

**Comment by october262** [comment]
does it not work with a VPN ??

**Comment by dirkf** [comment]
Possibly, but ATM I'll leave this to people able to access it directly. The question is whether OP's failure with cbs:... is repeatable given https://github.com/ytdl-org/youtube-dl/issues/31864#issuecomment-1473757202.

We now have ParamountPlus.com/gb but if it's anything like the offerings from the same group it'll be DRMed. Difficult to tell as even their device help page is broken,

**Comment by 4cq2** [comment]
> Somewhat surprisingly no yt-dlp user has reported this. Could it be an artefact of the access network, eg if not in region or if using a blocked VPN?

If `can.cbs.com` means Canada, then it would make sense to try Canada VPN. I retried the PHP URL with United States and Canada, still failing.

> and this test link https://www.paramountplus.com/movies/video/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_/ says it's DRM protected.

Several videos offer both DRM and non DRM streams. For example, the one you've commented on, both `DASH_CENC` (DRM) and `StreamPack` (not DRM) are offered. However in addition to the PHP error already discussed, another one has recently occured. If you make a `StreamPack` request, such as:

~~~
curl -v link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_
~~~

You get an HLS URL, as expected:

<https://cbsios-vh.akamaihd.net/i/temp_hd_gallery_video/CBS_Production_Outlet_VMS/video_robot/CBS_Production_Entertainment/2020/05/07/1735196227871/0_0_3436402_ful01_2588_503000.mp4.csmil/master.m3u8>

but if you actually try to request the HLS, you get:

~~~
HTTP/1.0 400 Bad Request
~~~

so it appears that all `StreamPack` options are dead, either temporarily or permanent, with the cause being an accident or a purposeful action by CBS. So until something changes any `StreamPack` options are essentially DRM.

> the same video at paramount + https://www.paramountplus.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_/ works with yt-dlp & ytdl-patched but does not work with youtube-dl.

this is a `Downloadable` type, which might be the only non DRM type still available. Note that previously an `HLS_CLEAR` type was available as well, but I think those have been dropped, not sure.

> Someone (OP perhaps?) will have to do some research, eg on how this works in region in a browser, to fix this.

I plan to work on this some more this weekend. I want to see what a web client is doing on the few non DRM options left, and also check Android client. Also I have code that is able to download and decrypt the DRM options, but I assume that is still beyond the scope of these projects.

**Comment by 4cq2** [comment]
> the same video at paramount + https://www.paramountplus.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_/ works with yt-dlp & ytdl-patched but does not work with youtube-dl.
> 


also I just noticed, YT-DLP **does not** work with this. From my testing is shows DRM for both URL, even though I have already proved at least one URL has a non DRM option.

**Comment by forty875** [comment]
Does this endpoint no longer work can.cbs.com/thunder/player/videoPlayerService.php ? I get status code 200 but no output...

**Comment by dirkf** [comment]
From the issue text:
>looks like the issue is that these type URLs are now returning empty responses:
>
>http://can.cbs.com/thunder/player/videoPlayerService.php?partner=cbs&contentId=wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_

So, yes.

**Comment by forty875** [comment]
How to get all the pids for all asset types now for example DASH_CENC , PRECON or HLS ? link.theplatform.com needs pid to get  m3u or mpd manifest...

**Comment by 4cq2** [comment]
@forty875 thats the whole point of this issue. Currently you cannot get that information. possibly we could discover this by using an Apple web client (Safari), or impersonating one. Alternatively, if someone knows how to MITM iOS or iPadOS that should work as well

**Comment by dirkf** [comment]
Won't a trace of playing a test video in a desktop browser do?

**Comment by 4cq2** [comment]
@dirkf regarding web clients, if you not using Safari, or a Safari User-Agent, or another Apple specific browser (I dont know what those would be), then likely you are just going to be served DASH, either Widevine or clear. So using an Apple client, either desktop or mobile is an important point here, for any who might care about HLS formats. I didn't make time for this over the last couple of days (busy with other stuff), but I will try to look at it this week, unless someone beats me to it.

**Comment by ParadoxGBB** [comment]
For me the original XML parse error is a somewhat separate issue from whether they are DRM'd that cropped up this week after working fine --- I had been using yt-dlp to download the subtitles and thumbnails without the content via the --skip-download parameter. Started failking for this series Tuesday for me.

Examples:

"https://www.cbs.com/shows/video/yJDiL4UCGqhPMzYeI7_uGETDWFuGRNui/"
"https://www.cbs.com/shows/video/TTobd7NjENoNESGeCpmsirueC4HOYDge/"
"https://www.cbs.com/shows/video/mjADR3rryehJKHRKqapazoYY0UX_6_F_/"

It's also retroactive with previously working videos apparently now failing, so an unresponsive endpoint seems logical to me.

Hope folks can help figure something out.

**Comment by dirkf** [comment]
For the Colbert show, I get these interesting JS fragments in the non-JS page seen by yt-dl, after pretty-printing:
<details><summary>Colbert fragments</summary>

```py
player.metaData = {
  'airdate_iso': '2023-03-29T20:35:00-07:00',
  'airdate_tv': '2023-03-29',
  'assetType': 'DASH_CENC',
  'brand': 'CBS',
  'mediaType': 'Full Episode',
  'channelName': null,
  'contentId': 'mjADR3rryehJKHRKqapazoYY0UX_6_F_',
  'contentUrl': '/shows/video/mjADR3rryehJKHRKqapazoYY0UX_6_F_/',
  'endCreditsChapterTime': 2430,
  'episodeNumber': '94',
  'ESTURLs': {
    'amazon': '',
    'iTunes': ''
  },
  'excludeOztam': null,
  'fullEpisode': true,
  'isServiceAllowed': true,
  'oztamMediaId': null,
  'pid': 'Z8rFLBAAODE5',
  'daistreamKey': null,
  'previewImageURL': 'https://thumbnails.cbsig.net/CBS_Production_Entertainment_VMS/2023/03/05/2178777667682/colbert_1396_2141418_1920x1080.jpg',
  'rating': 'TV-14',
  'regionalRatings': [
    {
      'region': 'US',
      'rating': 'TV-14',
      'disclaimer': null,
      'secondaryDescriptors': null,
      'subratings': null,
      'consumerAdvice': null,
      'ratingIcon': null
    }
  ],
  'seasonNumber': '8',
  'seriesTitle': 'The Late Show with Stephen Colbert',
  'showPageURL': 'shows/the-late-show-with-stephen-colbert/video',
  'subscriptionLevel': 'FREE',
  'thumbnail': {
    'height': 1080,
    'width': 1920,
    'assetType': 'Thumbnail',
    'url': 'https://thumbnails.cbsig.net/CBS_Production_Entertainment_VMS/2023/03/05/2178777667682/colbert_1396_2141418_1920x1080.jpg'
  },
  'thumbnailSheet': {
    'height': 126,
    'width': 54656,
    'assetType': 'ThumbSheet',
    'url': 'https://thumbnails.cbsig.net/CBS_Production_Entertainment_VMS/2023/03/05/2178777667682/COLBERT_1396_2997DF_1920X1080_178_2CH_PRORESHQ_2141419_vr_cs_224x126_10_244.jpg'
  },
  'tvRatingFlag': true,
  'videoLength': 2439,
  'videoPageURL': '/shows/video/mjADR3rryehJKHRKqapazoYY0UX_6_F_/',
  'videoTitle': 'The Late Show - 3/29/23 (Taron Egerton, Jay Chandrasekhar)',
  'label': '3/29/23 (Taron Egerton, Jay Chandrasekhar)',
  'videoProperties': [
    'Branded',
    'Current Season',
    'Episodeless'
  ],
  'playbackEvents': {
    'endCreditChapterTimeMs': 2430000,
    'previewStartTimeMs': null,
    'previewEndTimeMs': null,
    'openCreditEndTimeMs': null,
    'openCreditStartTime': null
  }
};
player.apiMetadata = {
  'genre': 'Talk',
  'status': 'AVAILABLE',
  '_firstIngestDate': '03/30/23 02:26 EDT',
  'expired': false,
  'showPageUrl': 'shows/the-late-show-with-stephen-colbert/video',
  'cbsShowId': 61456254,
  'primaryCategory': '518039619867',
  'primaryCategoryName': 'Late Night/Late Show/Colbert/Full Episodes',
  'editDate': 1680157555000,
  '_editDate': '03/30/23 02:25 EDT',
  '_lastModifiedDate': '03/30/23 02:26 EDT',
  'ingestDate': 1680157561000,
  'airDate': 1680147300000,
  'description': 'Stephen welcomes actor Taron Egerton and comedian Jay Chadrasekhar.',
  'shortDescription': 'Stephen welcomes actor Taron Egerton and comedian Jay Chadrasekhar.',
  'pubDate': 1680157561000,
  'label': '3/29/23 (Taron Egerton, Jay Chandrasekhar)',
  'videoPageUrl': 'shows/the-late-show-with-stephen-colbert/video/mjADR3rryehJKHRKqapazoYY0UX_6_F_/the-late-show-3-29-23-taron-egerton-jay-chandrasekhar-',
  'url': 'shows/the-late-show-with-stephen-colbert/video/mjADR3rryehJKHRKqapazoYY0UX_6_F_/the-late-show-3-29-23-taron-egerton-jay-chandrasekhar-',
  'assetType': 'DASH_CENC',
  'firstIngestDate': 1680157561000,
  'category': '',
  'topLevelCategory': 'Entertainment',
  'fullEpisode': true,
  'exclusive': false,
  'contentId': 'mjADR3rryehJKHRKqapazoYY0UX_6_F_',
  'title': 'The Late Show - 3/29/23 (Taron Egerton, Jay Chandrasekhar)',
  '_ingestDate': '03/30/23 02:25 EDT',
  'episodeNum': '94',
  '_pubDate': '03/30/23 02:26 EDT',
  'seasonNum': '8',
  'brand': 'CBS',
  'pid': 'Z8rFLBAAODE5',
  'seriesTitle': 'The Late Show with Stephen Colbert',
  '_airDate': '03/29/23 23:35 EDT',
  'duration': 2439,
  'lastModifiedDate': 1680157561000,
  'rating': 'tv-14',
  'deviceType': 'PC',
  'thumbnail': 'https://thumbnails.cbsig.net/CBS_Production_Entertainment_VMS/2023/03/05/2178777667682/colbert_1396_2141418_1920x1080.jpg',
  'amazonESTURL': '',
  'itunesESTURL': '',
  'streamingUrl': 'https://vod-gcs-cedexis.cbsaavideo.com/intl_vms/2023/03/05/2178777667682/2141560_cenc_dash/stream.mpd',
  'expirationDate': 1680766500000,
  '_expirationDate': '04/06/23 03:35 EDT',
  '_pubDateISO': '2023-03-29T23:26:01-07:00',
  '_airDateISO': '2023-03-29T20:35:00-07:00',
  'playerLocUrl': 'http://can.cbs.com/thunder/player/chrome/canplayer.swf?allowScriptAccess=always&allowFullScreen=true&pid=Z8rFLBAAODE5&partner=cbs_us_free_desktop&autoPlayVid=false&owner=CBS Production Entertainment',
  'subscriptionLevel': 'FREE',
  'mediaAvailableDate': '2023-03-30T00:35:00-0700',
  'mediaAvailableDateEpoch': 1680161700000,
  'closedCaptionUrl': 'https://vod-gcs-cedexis.cbsaavideo.com/intl_vms/caption-rmp/COLBERT_1396_2997_1680155824.xml',
  'isLive': false,
  'cmsAccountId': 'dJ5BDC',
  'appleWatchListShowKey': 'SHOW_61456254_The-Late-Show-with-Stephen-Colbert',
  'isProtected': true,
  'mediaType': 'Full Episode',
  'excludeNielsenTracking': false,
  'endCreditsChapterTime': '40:30',
  'thumbnailSet': [
    {
      'height': 1080,
      'width': 1920,
      'assetType': 'Thumbnail',
      'url': 'https://thumbnails.cbsig.net/CBS_Production_Entertainment_VMS/2023/03/05/2178777667682/colbert_1396_2141418_1920x1080.jpg'
    }
  ],
  'downloadCountrySet': [
  ],
  'regionalRatings': [
    {
      'region': 'US',
      'rating': 'TV-14',
      'disclaimer': null,
      'secondaryDescriptors': null,
      'subratings': null,
      'consumerAdvice': null,
      'ratingIcon': null
    }
  ],
  'premiumFeatures': [
    ''
  ],
  'videoProperties': [
    'Branded',
    'Current Season',
    'Episodeless'
  ],
  'availableForProfileTypes': [
    'ADULT'
  ],
  'playbackEvents': {
    'endCreditChapterTimeMs': 2430000,
    'previewStartTimeMs': null,
    'previewEndTimeMs': null,
    'openCreditEndTimeMs': null,
    'openCreditStartTime': null
  },
  'embeddable': 'false',
  'copyright': 'CBS, Inc.',
  'addOns': [
  ],
  'isContentAccessibleInCAN': true,
  'brandSlug': 'cbs',
  'tmsprogramID': 'EP019062761435',
  'tmsseriesID': 'SH019062760000',
  'thumbnailSheetSet': [
    {
      'height': 126,
      'width': 54656,
      'assetType': 'ThumbSheet',
      'url': 'https://thumbnails.cbsig.net/CBS_Production_Entertainment_VMS/2023/03/05/2178777667682/COLBERT_1396_2997DF_1920X1080_178_2CH_PRORESHQ_2141419_vr_cs_224x126_10_244.jpg'
    }
  ],
  '__FOR_TRACKING_ONLY_MEDIA_ID': 2178777667707
};
player.tms_program_id = 'EP019062761435';
// Set player.drm here/inline unless promise for client-side AJAX
if (!window.CBS.Registry.drmPromise) {
  player.drm = {
    'enabled': true,
    'sessionId': '',
    'ticket': '',
    'name': 'Widevine',
    'widevine': {
      'url': 'https://cbsi.live.ott.irdeto.com/widevine/getlicense?CrmId=cbsi&AccountId=cbsi&SubContentType=Default&ContentId=mjADR3rryehJKHRKqapazoYY0UX_6_F_',
      'header': {
        'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsImtpZCI6IjNkNjg4NGJmLWViMDktNDA1Zi1hOWZjLWU0NGE1NmY3NjZiNiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhbm9ueW1vdXNfVVMiLCJlbnQiOlt7ImJpZCI6IkFsbEFjY2Vzc01haW4iLCJlcGlkIjo3fV0sImlhdCI6MTY4MDE5NTgzNSwiZXhwIjoxNjgwMjAzMDM1LCJpc3MiOiJjYnMiLCJhaWQiOiJjYnNpIiwiaXNlIjp0cnVlLCJqdGkiOiJjZjhhZDc3NS03YTA3LTQyY2YtYjM2OC03M2NhZDM1ZDMyMWEifQ.OGbUVBCX5e1oVsDbbpglZuqUCxgjS7kWvz9f0w0lTh0'
      }
    }
  };
```

</summary>
Looks like no Colbert for yt-dl.

**Comment by ParadoxGBB** [comment]
No, but as the closed captioning jpeg and thumbnail are correctly published it seems I would have expected yt-dlp to move on from the XML error with --skip-download and grab those. But I'm largely out of my depth here.

**Comment by 4cq2** [comment]
I am still quite busy with other stuff (just finished a new binary to text encoding and an HTTP package this week), but I will see if I can make some time this weekend. I wanted to work on my ProtoBuf package, but I suppose this is more important. I did try to reverse engineer the app secret with no luck. Its only 64 bit, so looks like it might just be a non cryptographic hash, but I tried CRC64, FNV and MapHash with no luck. If anyone knows any other common 64 bit hashes I can try those. Thanks

**Comment by 4cq2** [comment]
# April 2023

apparently the situation is even worse now. For this public URL:

<https://paramountplus.com/movies/video/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_>

these command:

~~~
curl -v link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra
curl -v link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra?format=preview
curl -v https://link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra
curl -v https://link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra?format=preview
~~~

all return:

~~~
HTTP/1.1 404 Not Found
~~~

I will keep working on this, as even my own tool is now broken.

**Comment by 4cq2** [comment]
apparently the situation is even worse now. For this public URL:

<https://paramountplus.com/movies/video/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_>

this URL fails:

<http://link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra?format=preview>

but for this public URL:

https://paramountplus.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_

the same URL works:

http://link.theplatform.com/s/dJ5BDC/media/guid/2198311517/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_?format=preview

**Comment by 4cq2** [comment]
# April 2023

OK from my comment here:

https://github.com/ytdl-org/youtube-dl/issues/31864#issuecomment-1474345789

I have this command:

~~~
curl -v link.theplatform.com/s/dJ5BDC/media/guid/2198311517/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_
~~~

which previously returned an HLS URL:

<https://cbsios-vh.akamaihd.net/i/temp_hd_gallery_video/CBS_Production_Outlet_VMS/video_robot/CBS_Production_Entertainment/2020/05/07/1735196227871/0_0_3436402_ful01_2588_503000.mp4.csmil/master.m3u8>

but now does not:

~~~json
{
   "title": "No AssetType/ProtectionScheme/Format Matches",
   "description": "None of the available releases match the specified AssetType, ProtectionScheme, and/or Format preferences",
   "isException": true,
   "exception": "NoAssetTypeFormatMatches",
   "responseCode": "412"
}
~~~

and this is still a valid ID:

<https://paramountplus.com/movies/video/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_>

**Comment by 4cq2** [comment]
OK so continuing with this URL:

<https://paramountplus.com/movies/video/wQH9yE_y_Dt4ekDYm3yelhhY2KXvOra_>

based on this:

https://whatismybrowser.com/guides/the-latest-user-agent/safari

I changed my User-Agent to:

~~~
Mozilla/5.0 (Macintosh; Intel Mac OS X 13_3) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.4 Safari/605.1.15
~~~

If I then visit the Paramount+ page above, I get this:

> Sorry, this video is currently unavailable. Please try again.
>
> Need additional help? Check out our troubleshooting tips [here](//paramountplus.com/faq).
>
> error code: 5000

So it seems HLS is no longer an option for certain videos, making those specific videos DRM only. Note the `Downloadable` videos are still available without DRM, for example:

<https://paramountplus.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_>

**Comment by october262** [comment]
take the Halo TV series for instance -https://www.paramountplus.com/shows/halo/ , none of the the season episodes will download, but any of the behind the scenes, special features & clips will download fine.

**Comment by 4cq2** [comment]
> take the Halo TV series for instance -https://www.paramountplus.com/shows/halo/ , none of the the season episodes will download, but any of the behind the scenes, special features & clips will download fine.

@october262 yes they will download, just not with the typical tools:

~~~
> paramount -b AmBcaCActcTer6wa7QjKWWzdtKrMlHBw -a 2
GET http://link.theplatform.com/s/dJ5BDC/media/guid/2198311517/AmBcaCActcTer6wa7QjKWWzdtKrMlHBw?format=preview
GET https://www.paramountplus.com/apps-api/v3.0/androidphone/irdeto-control/anonymous-session-token.json?at=ABAAAAAAAAAAAAAAAAAAAAAAj3mOj6dirS7jbI4bsY6aO5RdLzKfKzZpmuy1mHn%2BKFQ%3D
GET http://link.theplatform.com/s/dJ5BDC/media/guid/2198311517/AmBcaCActcTer6wa7QjKWWzdtKrMlHBw?assetTypes=DASH_CENC&formats=MPEG-DASH
GET https://vod-gcs-cedexis.cbsaavideo.com/intl_vms/2022/03/16/2013439043562/1368394_cenc_dash/Halo_101_Final_PayTV_SDR_ProResHQ_422_2398_3840x2160_200_51_LoRo_28Mar2022_FRMTD_PPlus_BCAST_en-US_1262472_3_aac_128/init.m4v
POST https://cbsi.live.ott.irdeto.com/widevine/getlicense?CrmId=cbsi&AccountId=cbsi&SubContentType=Default&contentId=AmBcaCActcTer6wa7QjKWWzdtKrMlHBw
1.99%   1.15 megabyte   1.12 megabyte/s
~~~

**Comment by 4cq2** [comment]
OK I think I made some progress. If you monitor the Android app, you see a request like this:

~~~
GET https://www.paramountplus.com/apps-api/v2.0/androidphone/video/cid/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_.json?at=ABB90oFocIWpreOn1w0QqC%2FEi1lZTogArlwKOuD7ZmbjVsg7oCYtHqHGM1UFjJU8dW0%3D HTTP/2.0
cookie: CBS_COM=N0EwMjY0MDVENTU3MzJCNzJBMEQzMkIyMDQ0MjQyQUU6MTcxMTkxODg2NTMyND...
~~~

In the response is this:

~~~
"playerLocUrl": "http://can.cbs.com/thunder/player/chrome/canplayer.swf?allowScriptAccess=always&allowFullScreen=true&pid=u4uKImT6OqQR&partner=pp_us_lcp_android&autoPlayVid=false&owner=CBS Production News",
~~~

that URL is actually invalid because of the spaces, here is correct:

<http://can.cbs.com/thunder/player/chrome/canplayer.swf?allowScriptAccess=always&allowFullScreen=true&pid=u4uKImT6OqQR&partner=pp_us_lcp_android&autoPlayVid=false&owner=CBS%20Production%20News>

If you extract that file, you get another file `canplayer~.swf`. If you extract that file, the largest file should be named something like `15.82`. Inside that file is some interesting stuff:

~~~
http://can.cbs.com/thunder/player/videoPlayerService.php?pid=$PID$&domain=$DOMAIN$&partner=$PARTNER$
http://can.cbs.com/thunder/player/videoPlayerService.php?pid=$PID$&domain=$DOMAIN$&partner=$PARTNER$&auth=$AUTH$
~~~

thats as far as I have gotten so far.

**Comment by dirkf** [comment]
In the Colbert JSON quoted above:
>  'playerLocUrl': 'http://can.cbs.com/thunder/player/chrome/canplayer.swf?allowScriptAccess=always&allowFullScreen=true&pid=Z8rFLBAAODE5&partner=cbs_us_free_desktop&autoPlayVid=false&owner=CBS Production Entertainment',

This just gives an opaque binary file with initial bytes `CWS`. Apparently it's a [compressed SWF](https://gist.github.com/jlhg/6377082) which can then be edited with a suitable SWF tool imported from prehistory, or even just a hex editor, but appears to be generic.

**Comment by 4cq2** [comment]
If it helps I just used this to extract

https://github.com/mcmilk/7-Zip-zstd

**Comment by 4cq2** [comment]
OK I did some more research, I did not find any method to enumerate all `assetType` values:

https://github.com/4cq2/mech/blob/v1.2.8/paramount/assetType.md

So it seems for now some values are hidden, only available to people who already know they exist. `Downloadable` is still a valid value, so some videos are still available without DRM.

**Comment by dirkf** [comment]
Somehow the web and Android app manage to construct a playable URL, regardless of this (possibly requiring WV/FP). How does that happen? What do playable URLs look like?

**Comment by 4cq2** [comment]
> Somehow the web and Android app manage to construct a playable URL, regardless of this (possibly requiring WV/FP). How does that happen? What do playable URLs look like?

Not really sure what you are asking here. The web client always chooses DRM, even if non DRM is available. In fact, it will prevent you from playing at all if you disable DRM browser plugin.

Regarding the Android client, as mentioned in the research document in my previous comment, the Android client only returns DRM options, even if non DRM options are available.

So regarding "playable URLs", it doesnt really matter what methods the official clients use to general playable URLs (at least as far as YouTube-DL/YT-DLP are concerned), as all playable URLs generated by official clients are DRM enabled. Again to be clear, non DRM options are still available, but they are more limited than ever, and it appears no method is currently available to even enumerate what non DRM options might be available. I didnt any research into the `auth` query string key, so its possible that might be the fix. If I had to guess, I would say its the same as the `LS_Session` value, but I didn't check that. My previous research document was nearly exhaustive, so I am about mentally checked out on this matter.

**Comment by dirkf** [comment]
Thanks: you answered the question anyway!

Some other platforms have legacy implementations that can't be updated but have to continue playing content: perhaps these limited non-DRM options are provided for such a reason.

**Comment by 4cq2** [comment]
one note - currently I think the only non DRM `assetTypes` value is `Downloadable`. I did find some examples with `mediaType` of `Full Episode`:

https://paramountplus.com/shows/video/YxlqOUdP1zZaIs7FGXCaS1dJi7gGzxG_

but none with `mediaType` of `Movie`, so its possible that all `Movie` are now DRM only. I am not sure of anything I just said, so if someone can prove me wrong, I will be happy to hear it.

**Comment by Tetracerus** [comment]
For this video:
https://www.cbsnews.com/video/sunday-morning-full-episode-4-2-2023/

We just need to parse index.html to find the URL to master.m3u8 file.  Snippet:

Sunday Morning Full Episode 4/2","timestamp":1680440400000,"duration":3760,"durationLabel":"01:02:40","label":null,"images":{"sd":"https://assets1.cbsnewsstatic.com/hub/i/r/2023/04/02/a457a56d-498d-420b-9ba1-d91fab2ee07b/thumbnail/640x360/6bd395cc9a6be06149ad04014c1d9296/smjanepauley040223-1848688-640x360.jpg","hd":"https://assets3.cbsnewsstatic.com/hub/i/r/2023/04/02/a457a56d-498d-420b-9ba1-d91fab2ee07b/thumbnail/1280x720/58ec9371cb82766448dfed42a78b2182/smjanepauley040223-1848688-640x360.jpg"},"previewUrl":"https://splice.amlg.io/api/v2/video/A0u8tIz70VHrE_UwZAjPeMOAa54TH4T5/preview/","video":"https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/master.m3u8","video2":"https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/master.m3u8","format":"application/x-mpegURL","url":"https://www.cbsnews.com/video/sunday-morning-full-episode-4-2-2023/

If we fetch https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/master.m3u8 - it contains metadata on all the video and audio formats.  Snippet:

#EXT-X-STREAM-INF:BANDWIDTH=634827,AVERAGE-BANDWIDTH=516066,CODECS="avc1.640028,mp4a.40.2",RESOLUTION=640x360,FRAME-RATE=29.970,VIDEO-RANGE=SDR,AUDIO="audio_aac",CLOSED-CAPTIONS=NONE
0402_SUNMO_FULL_1_1848685_375/stream.m3u8

That SUNMO_FULL.../stream.m3u8 is the video playlist:

Video/audio links (VLC can play) are:
AAC Audio:  https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/0402_SUNMO_FULL_1_en-US_1848685_aac_128/stream.m3u8
640x360:  https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/0402_SUNMO_FULL_1_1848685_375/stream.m3u8
768x432:  https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/0402_SUNMO_FULL_1_1848685_750/stream.m3u8
960x540:  https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/0402_SUNMO_FULL_1_1848685_1500/stream.m3u8
1280x720: https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/0402_SUNMO_FULL_1_1848685_2100/stream.m3u8
HLS 1920x1080:  https://prod.vodvideo.cbsnews.com/cbsnews/vr/hls/2023/03/22/2185212995641/1848685_hls/0402_SUNMO_FULL_1_1848685_3000/stream.m3u8

**Comment by 4cq2** [comment]
@Tetracerus is now blocked, they wont be able to comment further. Please pay attention, this is about CBS and Paramount+, NOT CBS News. the difference is important. If you have further on topic comments, please contact me privately to be unblocked.
